common: leica-sep: terminate any possible characters sent to MCU

htho-lgs opened the pull request:
We can not guarantee, that a possible transition on TX line of the UART
connection was interpreted as the start of a command by MCU.
Send termination characters before sending a command in order to make
sure that MCU interprets intentionally sent characters only.

htho-lgs committed:
common: leica-sep: terminate any possible characters sent to MCU

We can not guarantee, that a possible transition on TX line of the UART
connection was interpreted as the start of a command by MCU.
Terminate anything that was sent before by sending termination
characters before the command.
With this we make sure that we do not extend anything that was sent
before with our command and MCU interprets `<chunk><cmd>`.

Signed-off-by: Thomas Haemmerle <thomas.haemmerle@leica-geosystems.com>

zandrey reviewed (changes requested):
@htho-lgs A bit more context is required here... From the commit message, I did not get what the actual problem is that this patch is trying to solve.

Is it that MCU erroneously reacts on the TX line state change? Or is it CPU that toggles TX at some points in time that makes MCU believe that there is a command sent?

Please extend the description here.

-- andrey

htho-lgs committed:
common: leica_sep: empty RX FIFO before sending command

If we have a full RX FIFO, we read out and interpret what was received
before we even sent the command. This can lead to missing characters of
the response to our command.

Empty the RX FIFO before sending our command in order to interpret only
packets, which was received after we sent the command.

Signed-off-by: Thomas Haemmerle <thomas.haemmerle@leica-geosystems.com>

htho-lgs committed:
common: leica_sep: ignore unexpected packets

While waiting for the response to our packet, MCU can also send other
packets.
Instead of treat this as error, we can just ignore those packets and
continue with waiting for the response.

Signed-off-by: Thomas Haemmerle <thomas.haemmerle@leica-geosystems.com>

htho-lgs committed:
common: leica_sep: do not return in the receive loop

We have a check for an error after the loop and can just break.

Signed-off-by: Thomas Haemmerle <thomas.haemmerle@leica-geosystems.com>

htho-lgs commented:
v2:
- explain the need of termination before sending a command (hopefully) better
- add additional commits for fixing SEP communication

max186 commented on common/leica_sep.c:
```diff
@@ -53,7 +53,7 @@ static int leica_sep_transfer(struct leica_sep_funcs *f, char *cmd,
 
 		if (size <= bytes_received)
-			return -1;
+			break;
```
Isn't it the same as before?
If we break, here https://github.com/hexagon-geo-surv/u-boot-imx/blob/b6bab515d01eb63e14e639fc49d6c3694e2be753/common/leica_sep.c#L96 will return -1, since the command is not yet received

htho-lgs replied:
yes - but we have only one exit point for errors

max186 commented on common/leica_sep.c:
```diff
@@ -59,24 +60,29 @@ static int leica_sep_transfer(struct leica_sep_funcs *f, char *cmd,
-
-			if (bytes_received >= 2 &&
-			    data[bytes_received - 2] == '"')
+			if (header_received) {
```
Does this mean that the "unexpected" packets will necessary have a header?
Otherwise we continue to read till the end of the timeout and return anyway -1, because `command_received` is still false.

htho-lgs replied:
Yes - they have a header (since it is SEP) and even if they wouldn't, we just set `bytes_received` to zero if `\r\n` is received and continue with waiting for our header
